Add ability to change advertise IP.

r0mant opened the pull request:
This PR implements ability to change advertise IP of a single-cluster node. This allows users implement scenarios where they want to deliver their pre-installed and pre-configured clusters as a full virtual appliance such as Amazon AMI or VMWare OVF/OVA.

### How it works

It introduces a new operation which I called "reconfigure" (potentially, it can also be used to reconfigure more stuff than just IP/hostname). It comes with its own plan/FSM/etc and is basically a lightweight install that runs on top of the existing cluster state and regenerates all configuration and secrets, then starts all services. Runtime/application phases are not executed b/c everything has been installed during original install (and is kinda the point of this feature).

The hard part was figuring out all the stuff that needs to be repaired on the node in order to accommodate the changed IP address. I've split those in two phases - "pre-cleanup" which is executed right away before reconfiguring anything, and "post-cleanup" which fixes stuff once we have Kubernetes. Things like removing lingering network interfaces, cleaning up old packages in both local and cluster states, recreating pods and so on. I tried to put proper descriptions on respective phase executors to explain what they do and why.

### Implementation notes

I tried to take as many shortcuts as I could to keep it simple. For example, since we're dealing with a single-node, there's no agent and everything runs locally. Also, since "reconfigure" is basically a lightweight "install" operation, it reuses a lot of the installer logic, plan generation and so on.

### Additional

Also introduced a couple of new commands:

```
$ gravity system stop [--disable]
$ gravity system start [--enable]
```

which stop/start our systemd services on the node. For convenience as they're needed for reconfiguration (see the included docs for more info).

a-palchikov commented on lib/install/reconfigure/plan_test.go:
```diff
@@ -0,0 +1,200 @@
+		ID: phases.PreCleanupPhase,
+		Phases: []storage.OperationPhase{
+			{
+				ID:   fmt.Sprintf("%v%v", phases.PreCleanupPhase, phases.NetworkPhase),
```
nit: `fmt.Sprint(phases.PreCleanupPhase, phases.NetworkPhase)`

a-palchikov commented on lib/install/reconfigure/plan_test.go:
```diff
@@ -0,0 +1,200 @@
+	storage.DeepComparePhases(c, storage.OperationPhase{
+		ID: installphases.ConfigurePhase,
+		Data: &storage.OperationPhaseData{
+			Install: &storage.InstallOperationData{},
```
Is nil value an option here?

a-palchikov commented on lib/install/reconfigure/planbuilder.go:
```diff
@@ -0,0 +1,130 @@
+// on the node to the plan.
+func (b *PlanBuilder) AddPreCleanupPhase(plan *storage.OperationPlan) {
+	plan.Phases = append(plan.Phases, storage.OperationPhase{
+		ID:          phases.PreCleanupPhase,
```
`PreCleanup` does not relay too much meaning. As far as I understand, the split is between things running prior to kubernetes being available and after  - what about splitting around kubernetes? Something like `PreKubernetes` and `PostKubernetes`?

a-palchikov commented on lib/install/reconfigure/state.go:
```diff
@@ -0,0 +1,116 @@
+func (s State) Server() (*storage.Server, error) {
+	servers := s.Cluster.ClusterState.Servers
+	if len(servers) != 1 {
+		return nil, trace.BadParameter("expected 1 server, got: %#v", servers)
```
nit: I think it's better to have a good formatter even though this looks like logic error which should not be user-facing but still it could help debugging when server list outputs only the relevant bits.

a-palchikov commented on lib/install/reconfigure/state.go:
```diff
@@ -0,0 +1,116 @@
+	// through if a 1-node cluster was expanded. Need some way to deduce the
+	// most recent cluster state based on the local state.
+	if len(state.Cluster.ClusterState.Servers) != 1 {
+		return trace.BadParameter("reconfiguration is only supported for single-node clusters, this one appears to have %v nodes",
```
```go
log.Field("state", state.Cluster.ClusterState.Servers).Warn("Reconfiguring multi-node cluster.")
return trace.BadParameter("only a single node cluster can be reconfigured") // other bits of the message are not important
```

a-palchikov commented on lib/install/suite.go:
```diff
@@ -0,0 +1,861 @@
+			AppPackage: appPackage.String(),
+			Provider:   schema.ProviderAWS,
+			DNSConfig:  s.dnsConfig,
+		})
```
`err` is dropped

a-palchikov commented on lib/install/suite.go:
```diff
@@ -0,0 +1,861 @@
+		AccountID:   account.ID,
+		ClusterName: s.cluster.Domain,
+		Admin:       true,
+	})
```
`err` is dropped

a-palchikov commented on lib/install/reconfigure/plan.go:
```diff
@@ -0,0 +1,96 @@
+	builder.AddChecksPhase(plan)
+	builder.AddConfigurePhase(plan)
+	builder.AddPullPhase(plan)
+	builder.AddMastersPhase(plan)
```
should check the error return

a-palchikov commented on tool/gravity/cli/reconfigure.go:
```diff
@@ -0,0 +1,152 @@
+		},
+	})
+	if utils.IsContextCancelledError(err) {
+		InstallerCleanup()
```
this should handle an error

a-palchikov commented on lib/system/environ/uninstall.go:
```diff
@@ -131,6 +157,57 @@ func DisableAgentServices(logger log.FieldLogger) error {
+		if err != nil {
+			log.WithError(err).Warn("Failed to stop service.")
+			errors = append(errors, err)
+		} else if disable {
```
this should be a separate function (same applies for start/enable)

a-palchikov commented on tool/gravity/cli/reconfigure.go:
```diff
@@ -0,0 +1,152 @@
+		return trace.NotFound("The cluster doesn't appear to be installed on this node.")
+	}
+	if err == nil {
+		return trace.NotFound(`The cluster appears to be running on this node. Please stop all services using "gravity system stop" prior to initiating the operation.`)
```
Why `NotFound` error here? I think `AlreadyExists` fits better (for the lack of `AlreadyRunning`).

a-palchikov commented on tool/gravity/cli/reconfigure.go:
```diff
@@ -0,0 +1,152 @@
+	"github.com/gravitational/trace"
+)
+
+const reconfigureMessage = `WARNING: This action will launch the operation of reconfiguring the cluster installed on this node to use a new advertise IP %v.
```
`Warning: this operation will update the advertise address of this cluster to %q - would you like to proceed?`

a-palchikov commented on tool/gravity/cli/register.go:
```diff
@@ -139,6 +139,12 @@ func RegisterCommands(app *kingpin.Application) *Application {
 	g.ResumeCmd.Force = g.ResumeCmd.Flag("force", "Force execution of specified phase.").Bool()
 	g.ResumeCmd.PhaseTimeout = g.ResumeCmd.Flag("timeout", "Phase execution timeout.").Default(defaults.PhaseTimeout).Hidden().Duration()
 
+	g.ReconfigureCmd.CmdClause = g.Command("reconfigure", `Reconfigure the advertise address used by the node. Supported for single-node clusters only. Cluster services must be stopped (e.g. using "gravity system stop") prior to initiating the operation.`)
```
Can the services be shut down automatically if the user confirms the operation?

a-palchikov commented on tool/gravity/cli/system.go:
```diff
@@ -558,6 +559,68 @@ func systemServiceStatus(env *localenv.LocalEnvironment, pkg loc.Locator, servic
 	return nil
 }
 
+const systemStopMessage = `WARNING: This action will stop all Gravity and Kubernetes services on the node.
```
Do we want to display warning banners for these commands? Could corresponding high-level commands do this automatically?

r0mant commented on lib/install/reconfigure/phases/cleanup.go:
```diff
@@ -0,0 +1,91 @@
+func NewCleanup(p fsm.ExecutorParams, operator ops.Operator, packages pack.PackageService, client *kubernetes.Clientset) (fsm.PhaseExecutor, error) {
+	switch {
+	case strings.HasPrefix(p.Phase.ID, PreCleanupPhase):
+		switch phaseID := strings.Replace(p.Phase.ID, PreCleanupPhase, "", 1); phaseID {
```
Yeah, flatter plan structure is probably better in this case, will update.

r0mant commented on lib/install/reconfigure/phases/packages.go:
```diff
@@ -0,0 +1,93 @@
+	LocalPackages pack.PackageService
+}
+
+// Execute removes old configuration & secret packages from the node.
```
So yeah, currently only config/secret packages are marked with this label - but the way I was thinking, we're removing everything for the old node so if there were any other packages marked it, we'd have to remove them too anyway.

r0mant commented on lib/install/reconfigure/phases/pods.go:
```diff
@@ -0,0 +1,95 @@
+// All Kubernetes pods have to be recreated when the cluster comes up after
+// reconfiguration to ensure that no old pods are lingering (Kubernetes may
+// get confused after the old node is gone and keep pods in terminating state
+// for a long time) and that they mount proper service tokens.
```
That's what I thought/hoped for too, but unfortunately it doesn't seem to be the case. In my testing, the Kubernetes would often get "confused" (for the lack of better word) and all the pods for the old node would get stuck in terminating state indefinitely (even after the node is gone) messing up scheduling for the new ones, so I had to resort to recreating all of them.

r0mant commented:
@a-palchikov @knisbet Think I've addressed all/most of your guys feedback and the feedback I got from the product team. See the latest commit if you're interested in the changes but here's the gist of stuff worth mentioning:

* UX changed a bit. There's no `gravity reconfigure` command anymore - instead, a user uses `gravity start --advertise-addr=xxx` command to start the node with a new IP. Under the hoods it still does the "reconfigure" operation. Also, the stop command is just `gravity stop` now (not a system subcommand).
* No more requirement to have the original installer present on the node.
* Fixed bad progress output that would duplicate timestamps.
* Fixed an issue with installer intermittently failing to connect to the process.
* Originally I used to use the "site-export" package to determine the state of the cluster in the absence of the running cluster, which is problematic for numerous reasons. Now I've changed this to instead have the process periodically sync its cluster state to the local backend and use that.

@knisbet - with IP address not being present on the node - this is actually checked for and if I give `--advertise-addr=1.1.1.1` on my machine, it correctly returns an error. If it still happens for you (or if I misunderstood the issue), might be worth investigating.

a-palchikov commented on tool/gravity/cli/reconfigure.go:
```diff
@@ -150,3 +139,33 @@ func newReconfigurator(ctx context.Context, config *install.Config, state *recon
+	// The cluster should be installed but not running.
+	err := localenv.DetectCluster(env)
+	if err != nil && trace.IsNotFound(err) {
+		return nil, trace.BadParameter("Cluster doesn't appear to be installed on this node.")
```
nit: `Cluster` and `node` don't mix well when reading this - how about `Gravity is not installed`?

a-palchikov commented on tool/gravity/cli/register.go:
```diff
@@ -139,11 +139,16 @@ func RegisterCommands(app *kingpin.Application) *Application {
-	g.ReconfigureCmd.AdvertiseAddr = g.ReconfigureCmd.Flag("advertise-addr", "New advertise address the node will use. Must be present on the node. Will be auto-selected if not specified.").String()
-	g.ReconfigureCmd.FromService = g.ReconfigureCmd.Flag("from-service", "Run in service mode.").Hidden().Bool()
-	g.ReconfigureCmd.Confirm = g.ReconfigureCmd.Flag("confirm", "Suppress confirmation prompt.").Bool()
+	// TODO(r0mant): Add flags --planet and --teleport to stop/start command.
```
Why do you need these flags? Are they supposed to control individual services outside the context of the reconfiguration operation?

r0mant replied:
This is an old TODO item, forgot to remove.

a-palchikov commented on tool/gravity/cli/register.go:
```diff
@@ -139,11 +139,16 @@ func RegisterCommands(app *kingpin.Application) *Application {
+	g.StopCmd.Disable = g.StopCmd.Flag("disable", "Disable systemd services as well.").Bool()
+
+	g.StartCmd.CmdClause = g.Command("start", "Start Gravity services on the node.")
+	g.StartCmd.AdvertiseAddr = g.StartCmd.Flag("advertise-addr", "New advertise address the node will use. Must be present on the node.").String()
```
Is this parameter validated?

r0mant replied:
Yes, same way as the advertise-addr parameter for gravity install.

a-palchikov commented on tool/gravity/cli/stop_start.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	if enable {
+		env.PrintStep("Gravity services have been enabled and started")
```
nit: I really feel there's no need to make user differentiate these states - can the services be started and disabled?

a-palchikov commented:
@r0mant Also, the corresponding enterprise PR seems to be missing.

r0mant commented on docs/6.x/cluster.md:
```diff
@@ -1480,39 +1481,34 @@ There are a few restrictions and assumptions about this procedure to keep in min
 
-At this point you can create a machine image (AMI/OVF/OVA/etc). If needed, restart the services on the node after the
-image has been created:
+At this point the machine's snapshot (AMI/OVF/OVA/etc) can be taken. The `--disable` flag can be provided to ensure that
```
Hm, I suppose we could always disable/enable.

r0mant commented on lib/process/process.go:
```diff
@@ -1412,6 +1485,8 @@ func (p *Process) initService(ctx context.Context) (err error) {
 		p.startService(p.runApplicationsSynchronizer)
 		p.startService(p.runNodeLabelsReconciler(client))
 
+		go p.runClusterStateSynchronizer(p.context)
```
Didn't even know we had this.

a-palchikov replied:
It also drops the error from the function so we don't know if it started.

benarent commented on docs/6.x/cluster.md:
```diff
@@ -1464,6 +1464,54 @@ The command can also output the images in the json or yaml format which can come
+## Changing Node Advertise Address
+
+!!! note "Supported version":
+    Changing a node's advertise address is supported starting from Gravity 7.
```
I think we should fork our docs and put this into 7.0 vs putting it into 6.x. Unless this feature is party back-ported to an earlier version.

r0mant replied:
Yeah, we should. Not as a part of this PR though.

benarent commented on docs/6.x/cluster.md:
````diff
@@ -1464,6 +1464,54 @@ The command can also output the images in the json or yaml format which can come
+Once the image has been deployed on a new node, start the cluster providing a new advertise address configuration:
+
+```bash
+$ sudo gravity start --advertise-addr=<new-ip>
````
Do we support IPv4 and IPv6?

r0mant replied:
Only IPv4.
